### DISCUSSION TITLE
Using java package in mirth

### DISCUSSION THREAD

**Discussion by rutledj** [nextgenhealthcare/connect#5009] [Q&A]
What is the proper way to use the Arrays class in Mirth?

I tried importing using this:  importPackage(java.util.Arrays); but that gives an error

I tried just calling it  with below code:

```
var arr1 = ["John","Jane"];
var arr2 = ["Jane","John"];

logger.info('Arrays equal = ' + java.util.Arrays.equals(arr1,arr2).toString());`
```
I'm sure this should be simple but I'm not getting it to work.

**Comment by jonbartels** [comment]
Hmmm. That code looks right. A similar example might be this link which uses Array List: https://forums.mirthproject.io/forum/mirth-connect/support/175691-multiple-row-request-jdbc-to-one-pdf?p=175723#post175723

In your case `arr1` and `arr2` are JavaScript arrays. I wonder if Rhino is having issues converting those into Java arrays. 

Do you get any errors? What output do you get?

Per the javadoc - https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#equals-java.lang.Object:A-java.lang.Object:A-

> Returns true if the two specified arrays of Objects are equal to one another. The two arrays are considered equal if both arrays contain the same number of elements, and all corresponding pairs of elements in the two arrays are equal. Two objects e1 and e2 are considered equal if (e1==null ? e2==null : e1.equals(e2)). **In other words, the two arrays are equal if they contain the same elements in the same order**. Also, two array references are considered equal if both are null.

Order matters. Are you expecting an order-sensitive comparison?

**Comment by rutledj** [comment]
I'm really just trying to see if I can use the Arrays class methods in my channel. So, just expecting it to return a true or false I guess.  I get this error when it runs:

```
2022-02-09 10:26:28,196]  ERROR  (com.mirth.connect.connectors.js.JavaScriptReceiver:135): Error executing JavaScript Reader script ffddf20f-23f7-44fd-8744-d549197d18f9.
com.mirth.connect.server.util.javascript.JavaScriptExecutorException: com.mirth.connect.server.MirthJavascriptTransformerException: CHANNEL:
	array testingCONNECTOR:
	SourceSOURCE CODE:
	639: 640: 641: var arr1 = [0,1,4,6,8];642: var arr2 = [0,1,4,6,8];643: 644: logger.info('Arrays equal = ' + java.util.Arrays.equals(arr1,arr2).toString()); 645: }LINE NUMBER:
	644DETAILS:
	The choice of Java method java.util.Arrays.equals matching JavaScript argument types (object,object) is ambiguous; candidate methods are:     boolean equals(byte[],byte[])    boolean equals(boolean[],boolean[])    boolean equals(double[],double[])    boolean equals(float[],float[])    boolean equals(java.lang.Object[],java.lang.Object[])    boolean equals(short[],short[])    boolean equals(int[],int[])    boolean equals(long[],long[])    boolean equals(char[],char[])
	at ffddf20f-23f7-44fd-8744-d549197d18f9:644 (doScript)
	at ffddf20f-23f7-44fd-8744-d549197d18f9:646
	at com.mirth.connect.server.util.javascript.JavaScriptUtil.execute(JavaScriptUtil.java:75)
	at com.mirth.connect.connectors.js.JavaScriptReceiver.poll(JavaScriptReceiver.java:101)
	at com.mirth.connect.donkey.server.channel.PollConnectorJob.execute(PollConnectorJob.java:49)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:557)Caused by: com.mirth.connect.server.MirthJavascriptTransformerException: CHANNEL:
	array testingCONNECTOR:
	SourceSOURCE CODE:
	639: 640: 641: var arr1 = [0,1,4,6,8];642: var arr2 = [0,1,4,6,8];643: 644: logger.info('Arrays equal = ' + java.util.Arrays.equals(arr1,arr2).toString()); 645: }LINE NUMBER:
	644DETAILS:
	The choice of Java method java.util.Arrays.equals matching JavaScript argument types (object,object) is ambiguous; candidate methods are:     boolean equals(byte[],byte[])    boolean equals(boolean[],boolean[])    boolean equals(double[],double[])    boolean equals(float[],float[])    boolean equals(java.lang.Object[],java.lang.Object[])    boolean equals(short[],short[])    boolean equals(int[],int[])    boolean equals(long[],long[])    boolean equals(char[],char[])
	at ffddf20f-23f7-44fd-8744-d549197d18f9:644 (doScript)
```

**Comment by rutledj** [comment]
How would I declare my array variables to be a Java array type?

**Reply by jonbartels** [reply]
Doing it directly in JS might be easier - https://flexiple.com/javascript-array-equality/

**Comment by rutledj** [comment]
Yeah, I was just trying to understand how to use some of the Java classes in Mirth.

**Comment by tonygermano** [comment]
You picked a bad example to just try out some java classes. Javascript arrays are not Java arrays. It needs to figure out how to convert them.

I probably wouldn't bother with `java.util.Arrays.equals` unless you are comparing two Java arrays. Even then, Rhino does a little magic to let you call javascript `Array.prototype` methods on java arrays, so the javascript function in the link posted by @jonbartels should still work on Java arrays.

If you really want to call the function anyway, you can do it like this:
```js
java.util.Arrays['equals(java.lang.Object[],java.lang.Object[])'](arr1, arr2)
```

There is also no need to use `importPackage` in this situation (or most situations.) Specifying the fully-qualified class name removes the need to import.

**Comment by rutledj** [comment]
I wasn't primarily interested in the equals function, just picked it out of the blue to try using the class. Sounds like The Arrays class is best left alone unless there is a distinct need. Thanks for the info.

**Reply by jonbartels** [reply]
I think you'll find common examples on https://forums.mirthproject.io/ . More common Java classes that get used in MC might be:

* Apache StringUtils, FileUtils, and IOUtils - these are amazing, convenient libraries for common utility operations. I would die without `StringUtils.isNotBlank`
* Google Guava
* Joda Time or java.time libraries in more recent JDK versions
* JDBC calls
* Writing your own custom JAR then calling that from MC - It's a good way to write a Java library, unit test it in Java using JUnit, then call it from MC to do work

There is nothing wrong with using Java arrays in MC. As a learning exercise that list might help you find other examples to work from.
